### ISSUE TITLE
Stale GPS location is accepted

### ISSUE THREAD

**Issue by jace** [pnbruckner/ha-composite-tracker#32]
I've added a composite device using the trackers for the app, Unifi and nmap. Turns out if the app goes offline it is still counted as as being currently active and at home. I've confirmed this happens when:

- Phone is in aeroplane mode
- Phone has left the house and WiFi network, and there is no cloud integration to reach Home Assistant to update location

The composite tracker has `require_movement: true` but this has no effect as there are no updates to the location and the stale location is treated as still valid. The only fixes appear to be:

- Remove app from device list, or
- Ensure external URL is functioning (cloud or VPN access) for a location ping from outdoors

**Comment by pnbruckner** [comment]
It's possible that #28 would solve your problem. Unfortunately, I've been too busy on other things to finish testing and merge that change. Let me know what you think. I could try to spend some time on getting that merged.

**Comment by adwuk** [comment]
With the latest version of Android, this is more of a problem as apps need to be reopened before they will start monitoring, so if you forget, the app won't track you. We use OwnTracks which suffers from this.  At some point, if no updates are received, the composite tracker should mark the input entity as stale and ignore it until there are further updates.

An automation that makes use of the device_tracker.see service, and updates the stale device tracker location to not_home, in the event of WiFi/Bluetooth/BLE/Other monitoring detecting that a device has left home might help temporarily.

**Comment by pnbruckner** [comment]
> if no updates are received, the composite tracker should mark the input entity as stale and ignore it until there are further updates

HA entities are _always_ stale. That's just the nature of sampling systems. They are up to date (or as up to date as they'll ever be) at the moment they update, and then are immediately stale until the next instant of update. The composite tracker does not try to make a determination as to when an input entity is "too stale." Rather, it uses each update that appears to have newer data than the previous update it took (from any input.)

@jace I don't use "the app", so I'm not familiar with how it works and what attributes it has. Does it have any sort of "last seen" attribute that indicates when the current state was last updated?

Bottom line, the composite does not try (much) to make up for deficiencies in the input entities. If they make themselves seem up to date when they're not, that is an issue with the corresponding integration.

**Comment by adwuk** [comment]
> Rather, it uses each update that appears to have newer data than the previous update it took (from any input.)

I agree that all updates are immediately stale, however the composite tracker is prioritising the home/away state of GPS trackers over others regardless of how old the GPS updates are.  This is where the problem lies, as GPS trackers (at least in my experience) on Android phones can be unreliable due to an over zealous phone operating system.

I use a composite tracker with the Home Assistant app, OwnTracks and a number of binary sensors (WiFi connection, Bluetooth etc). Here are screenshots of the two GPS sensors, and it would appear that neither have a 'last seen' type of attribute.

Home Assistant App entity:
![Screenshot 2022-10-13 175406](https://user-images.githubusercontent.com/17555423/195659283-6e02ac94-9e44-4541-b57a-5f4c2b866b17.jpg)

OwnTracks entity:
![Screenshot 2022-10-13 175814](https://user-images.githubusercontent.com/17555423/195658998-e2f379da-cc54-4b76-840d-a0cfe1cf3ca9.jpg)

**Comment by adwuk** [comment]
Thanks. I will update to v2.3.0, test and report back.

**Comment by adwuk** [comment]
It doesn't seem to be behaving any differently, but maybe my YAML isn't passing the test.  Here is what I have in my configuration.yaml file:

 ```
 - platform: composite
    name: composite_tracker
    time_as: device_or_local
    require_movement: true
    entity_id:
      - entity: sensor.wifi_binary_state
        all_states: true
      - device_tracker.bluetooth_binary_state
      - device_tracker.owntracks_gps
      - device_tracker.companion_app_gps
```

The component starts with no errors, so have I interpreted the changes correctly?

**Comment by pnbruckner** [comment]
@adwuk I'm glad it's helping with your scenario. BTW, you could do the same with `device_tracker.bluetooth_binary_state` (since I'm assuming its `source_type` is not `gps` and that its state is only ever `home` or `not_home`):
```yaml
- platform: composite
  name: composite_tracker
  time_as: device_or_local
  require_movement: true
  entity_id:
    - entity: binary_sensor.wifi_binary_state
      all_states: true
    - entity: device_tracker.bluetooth_binary_state
      all_states: true
    - device_tracker.owntracks_gps
    - device_tracker.companion_app_gps
```

**Comment by pnbruckner** [comment]
@jace do you think the new feature will solve your problem as well?

**Comment by jace** [comment]
I've updated the composite tracker and tried adding `all_states`, but I get an error in the config validator:

```Error loading /config/configuration.yaml: mapping values are not allowed here```

Config:
```yaml
device_tracker:
  - platform: composite
    name: kiran
    time_as: local
    require_movement: true
    entity_id:
      - device_tracker.oneplus_9p_app
      - device_tracker.oneplus_9p_unifi
        all_states: true
      - device_tracker.oneplus_9p_nmap
        all_states: true
```

**Comment by jace** [comment]
Fixed and deployed. I think this will fix things for me, but I'll test to be sure.

I have one advanced use case that seems a little beyond this: sometimes I leave my phone at home but take a spare phone that I keep for contingencies.

The composite device in this config is a single device tracked through multiple integrations, and the "require movement" flag is made irrelevant by the network trackers, which do not track movement and will continue to see the device at home.

In effect, I need a higher level composite that combines both phones and only then applies the "require movement" condition to ignore the phone left behind at home – except the phone that is away is likely not reporting it's location anymore.

If the composite device tracker reports location _iff_ a location-aware tracker (like the HA app) is its primary, then I can simply make it part of another composite across devices, and problem solved.

**Comment by pnbruckner** [comment]
@jace why not create a second composite that is a composite of the main phone's composite and the secondary phone?
```yaml
device_tracker:
  - platform: composite
    name: kiran_oneplus_9p
    time_as: local
    entity_id:
      - device_tracker.oneplus_9p_app
      - entity: device_tracker.oneplus_9p_unifi
        all_states: true
      - entity: device_tracker.oneplus_9p_nmap
        all_states: true
  - platform: composite
    name: kiran
    time_as: local
    require_movement: true
    entity_id:
      - device_tracker.kiran_oneplus_9p
      - device_tracker.kiran_second_phone
```

**Comment by jace** [comment]
I setup a test automation for the device trackers. One of the trackers is constantly flipping overnight, likely due to the device going to sleep. I don't know the Jinja2 variable names to identify the entity. I guess this will need a longer timeout period for whichever upstream integration is responsible (app, Unifi or Nmap).

![Screenshot_2022-10-20-09-03-17-43_c3a231c25ed346e59462e84656a70e50.jpg](https://user-images.githubusercontent.com/26695/196850194-b9d3fb75-5c0d-4117-b813-34a72f10293e.jpg)

**Comment by pnbruckner** [comment]
@jace 

> I don't know the Jinja2 variable names to identify the entity.

Can you share the YAML code of your automation? I'd be happy to help you figure out the entity ID of the device tracker entity that is changing to not home.

Also, did you see my suggestion above about two layers of composite trackers? Does it make sense? Do you think that would work for you?

**Comment by jace** [comment]
The double layer composite tracker should work. I'll switch to it once I nail down the overactive home/away tracker. Here's the test automation:

```yaml
alias: People not at home
description: ""
trigger:
  - platform: state
    entity_id:
      - person.kiran_jonnalagadda
    from: home
  # Repeat for more users
condition: []
action:
  - service: notify.persistent_notification
    data:
      message: Someone is not home (who?)
      title: Someone is not home
mode: single
```

My intended use: turn on hallway lights when someone comes home between 6 PM and 6 AM. There's a separate automation to turn them off at 10 PM regardless of current state, but what's happening now is that the lights keep turning back on all night because one tracker is flaky.

**Comment by jace** [comment]
This the "lights on arrival" automation:

```yaml
alias: "Evening: Lights on arrival"
description: ""
trigger:
  - platform: state
    entity_id:
      - person.kiran_jonnalagadda
    to: home
  # More people here
condition:
  - condition: time
    after: "18:00:00"
    before: "22:00:00"
action:
  - service: scene.turn_on
    target:
      entity_id: scene.ambient_lights
    metadata: {}
  - type: turn_on
    device_id: 51560a88563a775b578750958ab18484
    entity_id: fan.air_purifier
    domain: fan
mode: single
```

**Comment by pnbruckner** [comment]
Try this:
```yaml
alias: People not at home
description: ""
trigger:
  - platform: state
    entity_id:
      - person.kiran_jonnalagadda
      - person.2
      - person.3
    from: home
condition: []
action:
  - service: notify.persistent_notification
    data:
      message: "{{ trigger.to_state.name }} is not home"
      title: Someone is not home
mode: single
```

> but what's happening now is that the lights keep turning back on all night because one tracker is flaky

But the automation has a time condition, so even if the trackers are flaky, the lights shouldn't be coming on all night, right???

**Comment by pnbruckner** [comment]
@jace can this issue be closed?

**Comment by jace** [comment]
Yes. I'm no longer relying on presence detection for my automations, so I'm not in a position to test carefully. This can be closed.
